paddle/operators/adam_op.cc

```diff
@@ -0,0 +1,100 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#include "paddle/operators/adam_op.h"
+
+namespace paddle {
+namespace operators {
+
+class AdamOp : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+ protected:
+  void InferShape(const framework::InferShapeContext &ctx) const override {
+    PADDLE_ENFORCE_NOT_NULL(ctx.InputVar("param"),
+                            "Input(param) of AdamOp should not be null.");
+    PADDLE_ENFORCE_NOT_NULL(ctx.InputVar("grad"),
+                            "Input(grad) of AdamOp should not be null.");
+    PADDLE_ENFORCE_NOT_NULL(ctx.InputVar("moment1"),
+                            "Input(moment1) of AdamOp should not be null.");
+    PADDLE_ENFORCE_NOT_NULL(ctx.InputVar("moment2"),
+                            "Input(moment2) of AdamOp should not be null.");
+
+    PADDLE_ENFORCE_NOT_NULL(ctx.OutputVar("param_out"),
+                            "Output(param_out) of AdamOp should not be null.");
+    PADDLE_ENFORCE_NOT_NULL(
+        ctx.OutputVar("moment1_out"),
+        "Output(moment1_out) of AdamOp should not be null.");
+    PADDLE_ENFORCE_NOT_NULL(
+        ctx.OutputVar("moment2_out"),
+        "Output(moment2_out) of AdamOp should not be null.");
+
+    PADDLE_ENFORCE_EQ(ctx.Input<Tensor>("param")->dims(),
+                      ctx.Input<Tensor>("grad")->dims(),
+                      "Two input of Adam Op's dimension must be same.");
+    PADDLE_ENFORCE_EQ(ctx.Input<Tensor>("grad")->dims(),
+                      ctx.Input<Tensor>("moment1")->dims(),
+                      "Two input of Adam Op's dimension must be same.");
+    PADDLE_ENFORCE_EQ(ctx.Input<Tensor>("moment1")->dims(),
+                      ctx.Input<Tensor>("moment2")->dims(),
+                      "Two input of Adam Op's dimension must be same.");
+
+    ctx.Output<framework::Tensor>("param_out")
+        ->Resize(ctx.Input<Tensor>("param")->dims());
+    ctx.Output<framework::Tensor>("moment1_out")
+        ->Resize(ctx.Input<Tensor>("moment1")->dims());
+    ctx.Output<framework::Tensor>("moment2_out")
+        ->Resize(ctx.Input<Tensor>("moment2")->dims());
+  }
+};
+
+class AdamOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  AdamOpMaker(framework::OpProto *proto, framework::OpAttrChecker *op_checker)
+      : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddInput("param", "input parameter");
+    AddInput("grad", "input gradient");
+    AddInput("moment1", "first moment");
```

Comment:
Also, we should be changing strings like `moment1`, `beta1` , etc to `moment_1` and `beta_1` because these are for the Python interface. In Python `_` is a separator for words in variable names.